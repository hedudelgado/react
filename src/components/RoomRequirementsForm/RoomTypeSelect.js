import React from 'react';
import { array, bool, object } from 'prop-types';
import Select from '../../layout/Select';

const RoomTypeSelect = ({
  dictionary,
  type,
  hasChildren,
  hasMultipleAdults,
  ...fieldProps
}) => (
  <Select {...fieldProps}>
    {
      type
        .filter(({ code }) => (hasChildren ? code === 'FAM' : code !== 'FAM'))
        .filter(({ code }) => (hasMultipleAdults ? code !== 'SB' : true))
        .filter(({ code }) => (!hasMultipleAdults ? code !== 'TWIN' : true))
        .map(({ code, dictionaryKey }) => (
          <option
            key={code}
            value={code}
          >
            {dictionary[dictionaryKey]}
          </option>
        ))
    }
  </Select>
);

RoomTypeSelect.propTypes = {
  type: array.isRequired,
  hasChildren: bool,
  hasMultipleAdults: bool,
  dictionary: object.isRequired,
};

export default RoomTypeSelect;
