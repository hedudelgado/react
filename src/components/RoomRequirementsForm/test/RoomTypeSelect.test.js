import React from 'react';
import { shallow } from 'enzyme';
import RoomTypeSelect from '../RoomTypeSelect';

describe('RoomTypeSelect Component', () => {
  let wrapper;

  const props = {
    id: 'test',
    name: 'test',
    label: 'label',
    touched: true,
    onChange: jest.fn(),
    onBlur: jest.fn(),
    dictionary: {
      'roomrequirements.cotrequired.true': 'Yes',
      'roomrequirements.cotrequired.false': 'No',
      'roomrequirements.type.double': 'Double',
      'roomrequirements.type.single': 'Single',
      'roomrequirements.type.accessible': 'Accessible',
      'roomrequirements.type.family': 'Family',
    },
    type: [{
      dictionaryKey: 'roomrequirements.type.double',
      code: 'DB',
    },
    {
      dictionaryKey: 'roomrequirements.type.twin',
      code: 'TWIN',
    },
    {
      dictionaryKey: 'roomrequirements.type.accessible',
      code: 'DIS',
    },
    {
      dictionaryKey: 'roomrequirements.type.family',
      code: 'FAM',
    },
    {
      dictionaryKey: 'roomrequirements.type.single',
      code: 'SB',
    }],
  };

  beforeEach(() => {
    wrapper = shallow(<RoomTypeSelect {...props} />);
  });

  describe('RoomTypeSelect Component Interface', () => {
    it('should render the component', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should only display Family type if 1 or more children has been selected', () => {
      const mockProps = {
        ...props,
        hasChildren: true,
        defaultValue: 'DB',
      };

      wrapper = shallow(<RoomTypeSelect {...mockProps} />);
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('option').at(0)).toHaveValue('FAM');
    });

    it('should display every room option except for Family if no children has been selected', () => {
      const newProps = {
        ...props,
        hasChildren: false,
        defaultValue: 'DB',
      };

      wrapper = shallow(<RoomTypeSelect {...newProps} />);
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('option').at(0)).not.toHaveValue('FAM');
    });

    it('should display every room option except for Single and Family if multiple adults and no children have been selected', () => {
      const newProps = {
        ...props,
        hasMultipleAdults: true,
        hasChildren: false,
        defaultValue: 'DB',
      };

      wrapper = shallow(<RoomTypeSelect {...newProps} />);
      expect(wrapper).toMatchSnapshot();
      expect(wrapper).not.toContainMatchingElement('option[value="FAM"]');
      expect(wrapper).not.toContainMatchingElement('option[value="SB"]');
    });

    it('should display every room option except for Family and Twin if one adults and no children have been selected', () => {
      const newProps = {
        ...props,
        hasMultipleAdults: false,
        hasChildren: false,
        defaultValue: 'DB',
      };

      wrapper = shallow(<RoomTypeSelect {...newProps} />);
      expect(wrapper).toMatchSnapshot();
      expect(wrapper).not.toContainMatchingElement('option[value="FAM"]');
      expect(wrapper).not.toContainMatchingElement('option[value="TWIN"]');
      expect(wrapper).toContainMatchingElement('option[value="SB"]');
    });
  });
});
