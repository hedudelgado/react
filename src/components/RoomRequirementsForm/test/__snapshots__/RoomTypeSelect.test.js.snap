// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomTypeSelect Component RoomTypeSelect Component Interface should display every room option except for Family and Twin if one adults and no children have been selected 1`] = `
<Select
  defaultValue="DB"
  halfWidth={false}
  hideIcon={false}
  hideMessage={false}
  id="test"
  label="label"
  name="test"
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
  placeholder=""
  touched={true}
>
  <option
    key="DB"
    value="DB"
  >
    Double
  </option>
  <option
    key="DIS"
    value="DIS"
  >
    Accessible
  </option>
  <option
    key="SB"
    value="SB"
  >
    Single
  </option>
</Select>
`;

exports[`RoomTypeSelect Component RoomTypeSelect Component Interface should display every room option except for Family if no children has been selected 1`] = `
<Select
  defaultValue="DB"
  halfWidth={false}
  hideIcon={false}
  hideMessage={false}
  id="test"
  label="label"
  name="test"
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
  placeholder=""
  touched={true}
>
  <option
    key="DB"
    value="DB"
  >
    Double
  </option>
  <option
    key="DIS"
    value="DIS"
  >
    Accessible
  </option>
  <option
    key="SB"
    value="SB"
  >
    Single
  </option>
</Select>
`;

exports[`RoomTypeSelect Component RoomTypeSelect Component Interface should display every room option except for Single and Family if multiple adults and no children have been selected 1`] = `
<Select
  defaultValue="DB"
  halfWidth={false}
  hideIcon={false}
  hideMessage={false}
  id="test"
  label="label"
  name="test"
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
  placeholder=""
  touched={true}
>
  <option
    key="DB"
    value="DB"
  >
    Double
  </option>
  <option
    key="TWIN"
    value="TWIN"
  />
  <option
    key="DIS"
    value="DIS"
  >
    Accessible
  </option>
</Select>
`;

exports[`RoomTypeSelect Component RoomTypeSelect Component Interface should only display Family type if 1 or more children has been selected 1`] = `
<Select
  defaultValue="DB"
  halfWidth={false}
  hideIcon={false}
  hideMessage={false}
  id="test"
  label="label"
  name="test"
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
  placeholder=""
  touched={true}
>
  <option
    key="FAM"
    value="FAM"
  >
    Family
  </option>
</Select>
`;

exports[`RoomTypeSelect Component RoomTypeSelect Component Interface should render the component 1`] = `
<Select
  halfWidth={false}
  hideIcon={false}
  hideMessage={false}
  id="test"
  label="label"
  name="test"
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
  placeholder=""
  touched={true}
>
  <option
    key="DB"
    value="DB"
  >
    Double
  </option>
  <option
    key="DIS"
    value="DIS"
  >
    Accessible
  </option>
  <option
    key="SB"
    value="SB"
  >
    Single
  </option>
</Select>
`;
